src/meta/src/hummock/manager/commit_epoch.rs

```diff
@@ -486,4 +312,143 @@ impl HummockManager {
             }
         }
     }
+
+    fn on_handle_add_new_table(
+        &self,
+        new_version_delta: &mut SingleDeltaTransaction<'_, '_>,
+        table_ids: &Vec<TableId>,
+        compaction_group_id: CompactionGroupId,
+        table_compaction_group_mapping: &mut HashMap<TableId, CompactionGroupId>,
+    ) -> Result<HashMap<TableId, CompactionGroupId>> {
+        if table_ids.is_empty() {
+            return Err(Error::CompactionGroup("empty table ids".to_string()));
+        }
+
+        let mut new_table_ids = HashMap::new();
+
+        for table_id in table_ids {
+            if let Some(info) = new_version_delta
+                .latest_version()
+                .state_table_info
+                .info()
+                .get(table_id)
+            {
+                return Err(Error::CompactionGroup(format!(
+                    "table {} already exist {:?}",
+                    table_id.table_id, info,
+                )));
+            }
+            table_compaction_group_mapping.insert(*table_id, compaction_group_id);
+            new_table_ids.insert(*table_id, compaction_group_id);
+        }
+
+        Ok(new_table_ids)
+    }
+
+    async fn correct_commit_ssts(
+        &self,
+        sstables: Vec<LocalSstableInfo>,
+        table_compaction_group_mapping: &HashMap<TableId, CompactionGroupId>,
+    ) -> Result<BTreeMap<u64, Vec<SstableInfo>>> {
+        let mut new_sst_id_number = 0;
+        let mut sst_to_cg_vec = Vec::with_capacity(sstables.len());
+        for commit_sst in sstables {
+            let mut group_table_ids: BTreeMap<u64, Vec<u32>> = BTreeMap::new();
+            for table_id in commit_sst.sst_info.get_table_ids() {
+                match table_compaction_group_mapping.get(&TableId::new(*table_id)) {
+                    Some(cg_id_from_meta) => {
+                        group_table_ids
+                            .entry(*cg_id_from_meta)
+                            .or_default()
+                            .push(*table_id);
+                    }
+                    None => {
+                        tracing::warn!(
+                            "table {} in SST {} doesn't belong to any compaction group",
+                            table_id,
+                            commit_sst.sst_info.get_object_id(),
+                        );
+                    }
+                }
+            }
+
+            new_sst_id_number += group_table_ids.len();
+            sst_to_cg_vec.push((commit_sst, group_table_ids));
+        }
+        let mut new_sst_id = next_sstable_object_id(&self.env, new_sst_id_number).await?;
+        let mut commit_sstables: BTreeMap<u64, Vec<SstableInfo>> = BTreeMap::new();
+
+        for (mut sst, group_table_ids) in sst_to_cg_vec {
+            for (group_id, _match_ids) in group_table_ids {
+                let branch_sst = split_sst(&mut sst.sst_info, &mut new_sst_id);
```

Comment:
For the newly generated sst it, should we also generate it by `next_sstable_object_id`?

Reply:
> For the newly generated sst

which line of code pls?

Reply:
Inside `split_sst`
```
pub fn split_sst(sst_info: &mut SstableInfo, new_sst_id: &mut u64) -> SstableInfo {
    let mut branch_table_info = sst_info.clone();
    branch_table_info.sst_id = *new_sst_id;
    sst_info.sst_id = *new_sst_id + 1;
    *new_sst_id += 1;

    branch_table_info
}
```

The new `sst_id` is just incrementing the in-memory state.

Reply:
Because `let mut new_sst_id = next_sstable_object_id(&self.env, new_sst_id_number).await?` has reserved exactly that much of SST IDs.

It's indeed obscure. Some comments and assertions can help a bit.